### PR TITLE
add benchmark script to profile pipeline APIs

### PR DESCRIPTION
**Pull request by lvhan028**
after #1507 
addressing issue #1494 
python >= 3.10 is recommended.

### CODE REVIEW

File: benchmark/profile_pipeline_api.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+                                                 gen_configs,
+                                                 do_preprocess=False):
+                session_id = output.session_id
+                n_token = output.generate_token_len
+                finish_reason = output.finish_reason
+                stats[session_id] = (n_token, finish_reason)
+                if finish_reason is not None:
+                    self.pbar.update(1)
+        else:
+            for output in self.pipe(prompts, gen_configs, do_preprocess=False):
```
Comment (grimoire) [inline benchmark/profile_pipeline_api.py:98]: hope we have progress bar here.

File: benchmark/profile_pipeline_api.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+            num_tokens_per_iter=args.num_tokens_per_iter,
+            max_prefill_iters=args.max_prefill_iters)
+    elif args.backend == 'pytorch':
+        engine_config = PytorchEngineConfig(
+            session_len=args.session_len,
+            cache_max_entry_count=args.cache_max_entry_count,
+            block_size=args.cache_block_seq_len,
+            max_batch_size=args.concurrency,
+            tp=args.tp,
+            thread_safe=True)
```
Comment (grimoire) [inline benchmark/profile_pipeline_api.py:236]: disable thread safe mode.

File: benchmark/profile_pipeline_api.py (hunk 1 above)
Comment (AllentDan) [reply benchmark/profile_pipeline_api.py:98]: progress bar can only be add inside `batch_infer`.

File: lmdeploy/serve/async_engine.py [hunk 3]
```diff
@@ -391,6 +392,7 @@ def batch_infer(
                 True, which means chat_template will be applied.
             adapter_name (str): the adapter name of slora for pytorch backend.
                 Pick one from adapters. Default to None, using the base model.
+            progress_bar (tqdm): the progress bar used in pipeline benchmark
```
Comment (AllentDan) [inline lmdeploy/serve/async_engine.py:395]: can we make it a true or false argument? May update `__call__` function as well.

Comment (lvhan028) [reply lmdeploy/serve/async_engine.py:395]: Sure. That's a good idea

Comment (lvhan028) [reply lmdeploy/serve/async_engine.py]: fixed

**Review by AllentDan — APPROVED** [review]
LGTM

**Review by grimoire — APPROVED** [review]

**Pull request merged**